GVFS/GVFS.Service/ProductUpgradeTimer.cs

```diff
@@ -53,70 +52,101 @@ public void Dispose()
             {
-                this.tracer.RelatedError(errorMessage);
+                metadata.Add("Exception", e.ToString());
             }
 
-            if (deleteExistingDownloads)
-            {
-                ProductUpgraderInfo.DeleteAllInstallerDownloads();
-            }
+            return metadata;
         }
 
-        private bool TryDownloadUpgrade(IProductUpgrader productUpgrader, out string errorMessage)
+        private void TimerCallback(object unusedState)
         {
+            string errorMessage = null;
+
             using (ITracer activity = this.tracer.StartActivity("Checking for product upgrades.", EventLevel.Informational))
             {
-                Version newerVersion = null;
-                string detailedError = null;
-                if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                try
                 {
-                    return false;
-                }
+                    // The upgrade check always goes against GitHub
+                    GitHubUpgrader productUpgrader = GitHubUpgrader.Create(
+                        this.tracer,
+                        false,
+                        false,
+                        out errorMessage);
 
-                if (!productUpgrader.TryQueryNewestVersion(out newerVersion, out detailedError))
-                {
-                    errorMessage = "Could not fetch new version info. " + detailedError;
-                    return false;
-                }
+                    if (productUpgrader == null)
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
 
-                if (newerVersion == null)
-                {
-                    // Already up-to-date
-                    // Make sure there a no asset installers remaining in the Downloads directory. This can happen if user
-                    // upgraded by manually downloading and running asset installers.
-                    ProductUpgraderInfo.DeleteAllInstallerDownloads();
-                    errorMessage = null;
-                    return true;
-                }
+                        return;
+                    }
 
-                if (productUpgrader.TryDownloadNewestVersion(out detailedError))
-                {
-                    errorMessage = null;
-                    return true;
+                    InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
+                    if (!prerunChecker.TryRunPreUpgradeChecks(out _))
+                    {
+                        return;
+                    }
+
+                    if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);

```

Comment:
If `UpgradeAllowed` returns `false`, is it always an error?  I am wondering if any of these should actually be `RelatedWarning`.

Reply:
I made most of these RelatedWarning